Changes for processing Studio Hour logs since the last successful processed date

anthonypos opened the pull request.

anthonypos committed:
- Updated MAD.Integration.Common nuget package, added PostConfigure method and cleaned up Program.cs.
- Added code to pass last processed date time to GetStudioHours method and the NameStudioHourQuery.sql file.

anthonypos committed:
Updated LogDate parameter name to LastProcessedDate

anthonypos committed:
Adding check for Error in StudioHourLog record

maitlandmarshall commented on MAD.Procore.StudioHoursUpload/Jobs/StudioHourUploadLogProducer.cs:
```diff
@@ -32,7 +31,12 @@ public async Task ProduceStudioHoursUploadLogs()
+                .FirstOrDefaultAsync(x => x.Region == this.procoreConfig.Name && x.ProcessedDate.HasValue && string.IsNullOrWhiteSpace(x.Error));
+
+            var studioHours = await this.studioHourClient.GetStudioHours(lastStudioLog?.ProcessedDate);
             var today = this.GetTodayUtc();
```
This will still only generate a log for today (GetTodayUtc)

It should loop from LastStudioLog's date to Today and generate the missing logs for all the studio hours. GetStudioHours will have to be called for each day in between LastStudioLog and Today, otherwise the office hours & man count will not reflect that particular day

maitlandmarshall commented on MAD.Procore.StudioHoursUpload/Jobs/StudioHourUploadLogProducer.cs:
```diff
@@ -32,7 +31,12 @@ public async Task ProduceStudioHoursUploadLogs()
+            //Retrieve last processed studio log and pass processed date time to GetStudioHours method
+            var lastStudioLog = await this.studioHourDbContext.StudioHourUploadLog
+                .OrderByDescending(x => x.ProcessedDate)
+                .FirstOrDefaultAsync(x => x.Region == this.procoreConfig.Name && x.ProcessedDate.HasValue && string.IsNullOrWhiteSpace(x.Error));
```
Remove the x.ProcessedDate.HasValue and string.IsNullOrWhiteSpace checks. If the log exists at all (processed / unprocessed /errored / not errored) then we do not want to generate it again

maitlandmarshall commented on MAD.Procore.StudioHoursUpload/MAD.Procore.StudioHoursUpload.csproj:
```diff
@@ -8,8 +8,8 @@
     <PackageReference Include="Dapper" Version="2.0.90" />
-    <PackageReference Include="MAD.API.Procore" Version="1.0.3" />
-    <PackageReference Include="MAD.Integration.Common" Version="1.1.14.3" />
+    <PackageReference Include="MAD.API.Procore" Version="2.0.1.2111030544" />
```
Please revert this back to v1.0.3, don't want to accidently introduce any breaking changes with v2 of the Procore lib

maitlandmarshall commented on MAD.Procore.StudioHoursUpload/Startup.cs:
```diff
@@ -7,12 +7,13 @@
 {
     internal class Startup
     {
+        private const string CronEveryXHours = "*/59 */23 * * *";
```
Remove this

maitlandmarshall commented on MAD.Procore.StudioHoursUpload/Startup.cs:
```diff
@@ -36,6 +37,16 @@ public void ConfigureServices(IServiceCollection serviceDescriptors)
+
+            var queueName = procoreConfig.Name.ToLower().Replace(" ", "_");
+
+            recurringJobManager.AddOrUpdate<StudioHourUploadLogProducer>($"{procoreConfig.Name}.{nameof(StudioHourUploadLogProducer)}.ProduceStudioHoursUploadLogs", y => y.ProduceStudioHoursUploadLogs(), CronEveryXHours, null, queueName);
```
Use named parameters and remove the CronEveryXHours. Just want to use the default CRON schedule which it defaults to if you pass through null.

maitlandmarshall commented on MAD.Procore.StudioHoursUpload/Queries/NamelyStudioHourQuery.sql:
```diff
@@ -13,7 +13,7 @@ AS (SELECT [Company email],
         WHERE LogDateTime =
         (
-            SELECT MAX(LogDateTime)FROM NamelyReportKeyValue
+            SELECT MIN(LogDateTime) FROM NamelyReportKeyValue WHERE LogDateTime >= ISNULL(@LastProcessedDate, '')
```
If no parameter is passed through, it should be MAX
No parameter = most up to date log value

anthonypos committed:
- Updated log producer to loop from the last studio log date to todays date and generate missing logs for the studio hours.
- Changed namely studio hours query to use MAX(LogDateTime) if a null last processed date value is passed to query.
- Reverted MAD.API.Procore version and recurring job registration code.

anthonypos committed:
Added code to ensure hte current date and time is used if the updated last processed date is in the future

maitlandmarshall commented on MAD.Procore.StudioHoursUpload/Jobs/StudioHourUploadLogProducer.cs:
```diff
@@ -32,15 +31,42 @@ public async Task ProduceStudioHoursUploadLogs()
             var today = this.GetTodayUtc();
+            var now = DateTime.UtcNow;
+
+            if (lastStudioLog?.ProcessedDate == null)
```
This should be if (lastStudioLog is null)
